### PR TITLE
[FEATURE] Cart functionality

### PR DESCRIPTION
**Pull request by turbo-ele**
added basic manager for sales-managers

added order-address to item, to be able to set address on each item

### CODE REVIEW

File: Api/Item.php [hunk 1]
```diff
@@ -53,11 +60,13 @@ public function getName()
 
     /**
      * @param $name
+     *
```
Comment (notsure) [inline Api/Item.php]: remove blank

File: Api/Item.php [hunk 2]
```diff
@@ -73,11 +82,13 @@ public function getNumber()
 
     /**
      * @param $number
+     *
```
Comment (notsure) [inline Api/Item.php]: remove all blanks

File: Api/Item.php [hunk 3]
```diff
@@ -53,11 +60,13 @@ public function getName()
 
     /**
      * @param $name
+     *
      * @return Item
      */
     public function setName($name)
     {
         $this->entity->setName($name);
+
```
Comment (notsure) [inline Api/Item.php:68]: blanks no good

File: Api/Item.php [hunk 4]
```diff
@@ -533,24 +572,124 @@ public function setProduct($product)
      * @return Product
      * @VirtualProperty
      * @SerializedName("product")
+     * @Groups({"cart"})
      */
     public function getProduct()
     {
-        if ($this->getEntity()->getProduct()) {
```
Comment (notsure) [inline Api/Item.php:539]: looks good but hope this change does not destroy another code

File: Api/Item.php [hunk 5]
```diff
@@ -533,24 +572,124 @@ public function setProduct($product)
...
+    {
+        $this->entity->setDeliveryAddress($deliveryAddress);
+
+        return $this;
+    }
+
+    /**
+     * Get deliveryAddress
+     *
+     * @return OrderAddress $deliveryAddress
```
Comment (notsure) [inline Api/Item.php:571]: shouldn't @ return be the last property

File: Api/Item.php [hunk 6]
```diff
@@ -533,24 +572,124 @@ public function setProduct($product)
...
+     * @VirtualProperty
+     * @SerializedName("deliveryAddress")
+     * @Groups({"cart"})
+     */
+    public function getDeliveryAddress()
+    {
+        $address = $this->entity->getDeliveryAddress();
+        if ($address) {
+            return new OrderAddress($address);
+        }
```
Comment (notsure) [inline Api/Item.php:581]: return null;

File: Api/Item.php [hunk 7]
```diff
@@ -533,24 +572,124 @@ public function setProduct($product)
...
     }
+
+    /**
+     * {@inheritDoc}
+     * TODO: default-price EUR?
+     */
+    public function getCalcPrice($quantity)
+    {
+        // TODO: if ($this->getUseProductsPrice()) {
+        if (($product = $this->getProduct())) {
```
Comment (notsure) [inline Api/Item.php]: remove variable declaration from if

File: Api/Item.php [hunk 8]
```diff
@@ -533,24 +572,124 @@ public function setProduct($product)
...
+    {
+        return $this->getQuantity();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getCalcPriceGroup()
+    {
+        if ($supplier = $this->getSupplier()) {
```
Comment (notsure) [inline Api/Item.php]: remove variable declaration from if

File: Api/Item.php [hunk 9]
```diff
@@ -533,24 +572,124 @@ public function setProduct($product)
...
+    {
+        return $this->currency;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getCalcPriceGroupContent()
+    {
+        if ($supplier = $this->getSupplier()) {
```
Comment (notsure) [inline Api/Item.php]: remove variable declaration from if

File: Api/Item.php [hunk 10]
```diff
@@ -533,24 +572,124 @@ public function setProduct($product)
...
+            return array(
+                'id' => $supplier->getId(),
+                'name' => $supplier->getName(),
```
Comment (notsure) [inline Api/Item.php:658]: remove comma

File: Api/Item.php [hunk 11]
```diff
@@ -176,6 +194,7 @@ public function getCreator()
     {
         // just return id
         $creator = $this->entity->getCreator();
+
         return !$creator ? null : array(
```
Comment (notsure) [inline Api/Item.php:188]: this is very very hard to read

tough it is valid

File: Entity/Item.php [hunk 12]
```diff
@@ -150,6 +156,7 @@ public function __construct()
      * Set name
      *
      * @param string $name
+     *
```
Comment (notsure) [inline Entity/Item.php]: empty lines

File: Item/ItemManager.php [hunk 13]
```diff
@@ -80,18 +88,22 @@ public function __construct(
...
-     * @internal param null $id
      * @return null|\Sulu\Bundle\Sales\CoreBundle\Api\Item
      */
-    public function save(array $data, $locale, $userId = null, Item $item = null, $itemStatusId = null)
+    public function save(array $data, $locale, $userId = null, $item = null, $itemStatusId = null)
     {
         // check requiresd data
-        $this->checkRequiredData($data, !!$item);
+        if (!$item) {
+            $this->checkRequiredData($data, !!$item);
```
Comment (notsure) [inline Item/ItemManager.php]: replace !!$item with "false" as this will always be false in this case

File: Item/ItemManager.php [hunk 14]
```diff
@@ -255,19 +267,23 @@ function (&$item) use ($locale) {
...
     private function checkRequiredData($data, $isNew)
     {
         // either name or products must be set
         if (array_key_exists('product', $data)) {
-            $this->checkDataSet($data['product'], 'id', $isNew);
+            // product-id must be defined
+            $this->getProductId($data['product']);
         } else {
-            $this->checkDataSet($data, 'name', $isNew);
+            $this->checkDataSet($data, 'name', !$isNew);
```
Comment (notsure) [inline Item/ItemManager.php]: why has this changed?

File: Item/ItemManager.php [hunk 15]
```diff
@@ -362,14 +402,16 @@ private function setItemByProductData($data, Item $item, $locale)
             $item->setTax(0);
 //            $item->setTax($product->getTaxClass()->getTax($locale));
 
+            // TODO: product-price
             if ($item->getUseProductsPrice() === true) {
-                $item->setPrice($product->getPrice());
+                $price = $this->productPriceManager->getBulkPriceForCurrency(
```
Comment (notsure) [inline Item/ItemManager.php:414]: getBulkPriceForCurrency 

doen't this function need a currency?

File: Manager/BaseSalesManager.php [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+     * sets a date if it's set in data
+     *
+     * @param $data
+     * @param $key
+     * @param $currentDate
+     * @param callable $setCallback
+     */
+    protected function setDate($data, $key, $currentDate, callable $setCallback)
+    {
+        if (($date = $this->getProperty($data, $key, $currentDate)) !== null) {
```
Comment (notsure) [inline Manager/BaseSalesManager.php]: remove variable declaration from if

File: Manager/BaseSalesManager.php [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+     * @param $key
+     * @param $data
+     *
+     * @return bool
+     */
+    protected function checkIfSet($key, $data)
+    {
+        $keyExists = array_key_exists($key, $data);
+
+        return $keyExists && $data[$key] !== null && $data[$key] !== '';
```
Comment (notsure) [inline Manager/BaseSalesManager.php:64]: use empty()
http://php.net/manual/en/function.empty.php

File: Manager/BaseSalesManager.php [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+            // TODO: inject repository class
+            $account = $this->em->getRepository(static::$accountEntityName)->find($accountData['id']);
+            if (!$account) {
+                // TODO custom exception: dependencyNotFound
+                throw new \Exception(sprintf("Entity '%s' with id %s not found", static::$accountEntityName, $accountData['id']));
+            }
+            $entity->setAccount($account);
+
+            return $account;
+        } else {
```
Comment (notsure) [inline Manager/BaseSalesManager.php]: else is not needed ?
as code will return anyway in the if cause above you could simple add:

$entity->setAccount(null);
return null;

File: Pricing/GroupedItemsPriceCalculator.php [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+    /**
+     * validate item values
+     *
+     * @param $item
+     *
+     * @throws PriceCalculationException
+     */
+    protected function validateItem($item)
+    {
+//        // item must be instance of PriceCalcualtionInterface
```
Comment (notsure) [inline Pricing/GroupedItemsPriceCalculator.php]: remove comments

File: Api/Item.php (hunk 2 above)
Comment (turbo-ele) [inline Api/Item.php]: ha ha

File: Item/ItemManager.php (hunk 15 above)
Comment (turbo-ele) [inline Item/ItemManager.php:414]: it has a fallback value. this needs to be changed in future, just added a todo

File: Manager/BaseSalesManager.php (hunk 17 above)
Comment (turbo-ele) [inline Manager/BaseSalesManager.php:64]: empty returns true for data['index'] = 0;

File: Api/Item.php [hunk 20]
```diff
@@ -553,4 +591,73 @@ private function getFormatter($locale)
...
+    /**
+     * {@inheritDoc}
+     */
+    public function getCalcPriceGroupContent()
+    {
+        $supplier = $this->getSupplier();
+        if ($supplier) {
+            return array(
+                'id' => $supplier->getId(),
+                'name' => $supplier->getName(),
```
Comment (notsure) [inline Api/Item.php:658]: remove comma

**Pull request merged**